experimental/form-engine/ARCHITECTURE.md

```diff
@@ -0,0 +1,324 @@
+- Features depend on BaseFeature abstraction
+- ValidationErrorHandler depends on strategy abstraction
+
+### DRY Principle
+
+**Eliminated Duplications:**
+- SchedulerService centralizes all async scheduling (~50 lines eliminated)
+- BaseFeature provides shared state management (~80 lines eliminated)
+- Helper utilities eliminate null-checking boilerplate (~70 lines eliminated)
+- ValidationErrorHandler replaces nested conditionals (~200 lines total eliminated)
+
+### KISS Principle
+
+**Simplified Complex Code:**
+- FeatureFactory replaces verbose feature instantiation
+- ValidationErrorHandler replaces 80+ lines of if/else chains with 20 lines
+- Helper utilities (`safeCall`, `safeGet`) simplify common operations
+- BaseFeature provides clear template for feature development
+
+### Key Principles
+
+- **Single source of truth** - FormEngine coordinates all state
+- **Immutable updates** - State changes create new objects
+- **Selective subscriptions** - Only subscribe to needed state with bubble support
+- **Efficient caching** - Map-based caching with automatic size management
+- **WeakMap contexts** - Automatic cleanup of event listeners
+- **Explicit service boundaries** - Each service independently mockable
+- **Factory-based creation** - Consistent feature initialization
+- **Strategy-based handling** - Extensible error processing
+
+### Events
+
+**Form-level Events:**
+- `INIT` - Form initialized
+- `RESET` - Form reset to initial values
+- `CHANGE` - Any field value changed
+- `VALUES` - Form values updated
+- `ERROR` - Validation error occurred
+- `VALIDATION` - Validation completed
+- `SUBMIT` - Form submission
+- `CONFIG_UPDATE` - Configuration changed
+
+**Field-level Events:**
+- `change:${path}` - Specific field value changed
+- `error:${path}` - Specific field error set
+- `errors:${path}` - Specific field errors updated
+
+**Event Features:**
+- Event handlers wrapped with error handling and context tracking for safe cleanup
+- **Bubble support**: Listeners can subscribe with `{ bubble: true }` to receive nested field events
+  - Example: `engine.on('change:orders', callback, null, { bubble: true })` fires when `orders[0].amount` changes
+  - Used by `useWatch(name, { bubble: true })` for efficient array/object watching
+  - Optimization: Only emits to bubble listeners when nested fields change via `hasListener(event, { onlyBubble: true })`
+
+### Scheduling (SchedulerService)
+
+**Scheduling Strategies:**
+- **Microtask** - High-priority async tasks (before next render)
+- **AnimationFrame** - Visual updates aligned with browser repaints
+- **Timeout** - Delayed execution with configurable delay
+- **Immediate** - Next tick execution (microtask or timeout fallback)
+
+**Usage:**
+```javascript
+// Before: Manual scheduling with fallbacks everywhere
+if (typeof queueMicrotask !== 'undefined') {
+  queueMicrotask(callback);
+} else {
+  Promise.resolve().then(callback);
+}
```

Comment:
Under what circumstances was `queueMicrotask` ever undefined previously? e.g. was this just a testing problem with jsdom?